File: server/src/intTest/java/org/tctalent/server/repository/db/SavedListRepositoryIntTest.java
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedSavedList;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedSavedSearch;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedSfJobOpp;
import static org.tctalent.server.repository.db.integrationhelp.DomainHelpers.getSavedUser;

import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.tctalent.server.model.db.SalesforceJobOpp;
import org.tctalent.server.model.db.SavedList;
import org.tctalent.server.model.db.SavedSearch;
import org.tctalent.server.model.db.Status;
import org.tctalent.server.model.db.User;
import org.tctalent.server.repository.db.integrationhelp.BaseDBIntegrationTest;

public class SavedListRepositoryIntTest extends BaseDBIntegrationTest {

  @Autowired
  private SavedListRepository repo;
  @Autowired
  private SalesforceJobOppRepository sfJobOppRepository;
  @Autowired
  private UserRepository userRepository;
  @Autowired
  private SavedSearchRepository savedSearchRepository;
  private SavedList savedList;
  private SalesforceJobOpp testSFJobOpp;
  private User testUser;
  private SavedSearch testSavedSearch;

  @BeforeEach
  public void setup() {
    assertTrue(isContainerInitialised());

    savedList = getSavedSavedList(repo);
    testSFJobOpp = getSavedSfJobOpp(sfJobOppRepository);
    testUser = getSavedUser(userRepository);
  }

  @Test
  public void testFindByJobIds() {
    savedList.setSfJobOpp(testSFJobOpp);
    repo.save(savedList);
    List<SavedList> result = repo.findByJobIds(testSFJobOpp.getId());
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(savedList.getId(), result.getFirst().getId());
  }

  @Test
  public void testFindByJobIdsFail() {
    List<SavedList> result = repo.findByJobIds(testSFJobOpp.getId());
    assertNotNull(result);
    assertTrue(result.isEmpty());
  }

  @Test
  public void testFindByNameIgnoreCase() {
    repo.save(savedList);
    SavedList results = repo.findByNameIgnoreCase(savedList.getName(),
        savedList.getCreatedBy().getId()).orElse(null);
    assertNotNull(results);
    assertEquals(savedList.getId(), results.getId());
  }

  @Test
  public void testFindByNameIgnoreCaseFail() {
    repo.save(savedList);
    SavedList results = repo.findByNameIgnoreCase("NONE", testUser.getId()).orElse(null);
    assertNull(results);
  }

  @Test
  public void testFindByIdLoadUsers() {
    SavedList result = repo.findByIdLoadUsers(savedList.getId()).orElse(null);
    assertNotNull(result);
    assertEquals(savedList.getId(), result.getId());
  }

  @Test
  public void testFindByIdLoadUsersFailNoId() {
    SavedList result = repo.findByIdLoadUsers(0L).orElse(null);
    assertNull(result);
  }

  @Test
  public void testFindByIdLoadCandidates() {
    SavedList result = repo.findByIdLoadCandidates(savedList.getId()).orElse(null);
    assertNotNull(result);
    assertEquals(savedList.getId(), result.getId());
  }

  @Test
  public void testFindByIdLoadCandidatesFailNoId() {
    SavedList result = repo.findByIdLoadCandidates(0L).orElse(null);
    assertNull(result);
  }

  @Test
  public void testFindSelectionList() {
    testSavedSearch = getSavedSavedSearch(savedSearchRepository);
    savedList.setSavedSearch(testSavedSearch);
    repo.save(savedList);
    SavedList result = repo.findSelectionList(testSavedSearch.getId(),
        savedList.getCreatedBy().getId()).orElse(null);
    assertNotNull(result);
    assertEquals(savedList.getId(), result.getId());
  }

  @Test
  public void testFindSelectionListFailNoSearch() {
    testSavedSearch = getSavedSavedSearch(savedSearchRepository);
    SavedList result = repo.findSelectionList(testSavedSearch.getId(),
        savedList.getCreatedBy().getId()).orElse(null);
    assertNull(result);
  }

  @Test
  public void testFindSelectionListFailUser() {
    testSavedSearch = getSavedSavedSearch(savedSearchRepository);
    SavedList result = repo.findSelectionList(testSavedSearch.getId(), null).orElse(null);
    assertNull(result);
  }

  @Test
  public void testFindRegisteredJobList() {
    savedList.setSfJobOpp(testSFJobOpp);
    repo.save(savedList);
    assert savedList.getSfJobOpp() != null;
    SavedList result = repo.findRegisteredJobList(savedList.getSfJobOpp().getSfId()).orElse(null);
    assertNotNull(result);
    assertEquals(savedList.getId(), result.getId());
  }

  @Test
  public void testFindRegisteredJobListFailNotRegistered() {
    repo.save(savedList);
    savedList.setRegisteredJob(false);
    SavedList result = repo.findRegisteredJobList(String.valueOf(testSFJobOpp.getId()))
        .orElse(null);
    assertNull(result);
  }

  @Test
  public void testFindRegisteredJobListFailId() {
    repo.save(savedList);
    // This is unusual to have a string for the sfJobLink
    SavedList result = repo.findRegisteredJobList("").orElse(null);
    assertNull(result);
  }

  @Test
  public void testFindByShortNameIgnoreCase() {
    SavedList result = repo.findByShortNameIgnoreCase(savedList.getTbbShortName()).orElse(null);
    assertNotNull(result);
    assertEquals(savedList.getId(), result.getId());
  }

  @Test
  public void testFindByShortNameIgnoreCaseFail() {
    SavedList result = repo.findByShortNameIgnoreCase("BOB").orElse(null);
    assertNull(result);
  }

  @Test
  public void testFindListsWithJobs() {
    savedList.setSfJobOpp(testSFJobOpp);
    repo.save(savedList);
    List<SavedList> result = repo.findListsWithJobs();
    assertNotNull(result);
    assertFalse(result.isEmpty());
    assertEquals(savedList.getId(), result.getFirst().getId());
  }

  @Test
  public void testFindListsWithJobsFail() {
    savedList.setStatus(Status.deleted);
    repo.save(savedList);
    List<SavedList> result = repo.findListsWithJobs();
    assertNotNull(result);
    assertTrue(result.isEmpty());
  }

  @Test
  public void testFindListsWithJobsFailJobOppNull() {
    List<SavedList> result = repo.findListsWithJobs();
    assertNotNull(result);
    assertTrue(result.isEmpty());
  }
}

File: server/src/intTest/kotlin/org/tctalent/server/repository/db/SavedListRepositoryIntTest.kt
/*
 * Copyright (c) 2024 Talent Beyond Boundaries.
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License
 * for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package org.tctalent.server.repository.db

import kotlin.jvm.optionals.getOrNull
import kotlin.test.*
import org.springframework.beans.factory.annotation.Autowired
import org.tctalent.server.model.db.*
import org.tctalent.server.repository.db.integrationhelp.*

class SavedListRepositoryIntTest : BaseDBIntegrationTest() {
  @Autowired lateinit var repo: SavedListRepository
  @Autowired private lateinit var sfJobOppRepository: SalesforceJobOppRepository
  @Autowired private lateinit var userRepository: UserRepository
  @Autowired private lateinit var savedSearchRepository: SavedSearchRepository
  private lateinit var savedList: SavedList
  private lateinit var testSFJobOpp: SalesforceJobOpp
  private lateinit var testUser: User
  private lateinit var testSavedSearch: SavedSearch

  @BeforeTest
  fun setup() {
    assertTrue { isContainerInitialized() }
    savedList = getSavedSavedList(repo)
    testSFJobOpp = getSavedSfJobOpp(sfJobOppRepository)
    testUser = getSavedUser(userRepository)
  }

  @Test
  fun `test find by job ids`() {
    repo.save(savedList.apply { sfJobOpp = testSFJobOpp })
    val result = repo.findByJobIds(testSFJobOpp.id)
    assertNotNull(result)
    assertTrue { result.isNotEmpty() }
    assertEquals(savedList.id, result.first().id)
  }

  @Test
  fun `test find by job ids fail`() {
    val result = repo.findByJobIds(testSFJobOpp.id)
    assertNotNull(result)
    assertTrue { result.isEmpty() }
  }

  @Test
  fun `test find by name ignore case`() {
    repo.save(savedList.apply { users = mutableSetOf(testUser) })
    val results = repo.findByNameIgnoreCase(savedList.name, savedList.createdBy.id).getOrNull()
    assertNotNull(results)
    assertEquals(savedList.id, results.id)
  }

  @Test
  fun `test find by name ignore case fail`() {
    repo.save(savedList.apply { users = mutableSetOf(testUser) })
    val results = repo.findByNameIgnoreCase("NONE", testUser.id).getOrNull()
    assertNull(results)
  }

  @Test
  fun `test find by id load users`() {
    val result = repo.findByIdLoadUsers(savedList.id).getOrNull()
    assertNotNull(result)
    assertEquals(savedList.id, result.id)
  }

  @Test
  fun `test find by id load users fail no id`() {
    val result = repo.findByIdLoadUsers(0L).getOrNull()
    assertNull(result)
  }

  @Test
  fun `test find by id load candidates`() {
    val result = repo.findByIdLoadCandidates(savedList.id).getOrNull()
    assertNotNull(result)
    assertEquals(savedList.id, result.id)
  }

  @Test
  fun `test find by id load candidates fail no id`() {
    val result = repo.findByIdLoadCandidates(0L).getOrNull()
    assertNull(result)
  }

  @Test
  fun `test find selection list`() {
    testSavedSearch = getSavedSavedSearch(savedSearchRepository)
    repo.save(savedList.apply { savedSearch = testSavedSearch })
    val result = repo.findSelectionList(testSavedSearch.id, savedList.createdBy.id).getOrNull()
    assertNotNull(result)
    assertEquals(savedList.id, result.id)
  }

  @Test
  fun `test find selection list fail no search`() {
    testSavedSearch = getSavedSavedSearch(savedSearchRepository)
    val result = repo.findSelectionList(testSavedSearch.id, savedList.createdBy.id).getOrNull()
    assertNull(result)
  }

  @Test
  fun `test find selection list fail user`() {
    testSavedSearch = getSavedSavedSearch(savedSearchRepository)
    val result = repo.findSelectionList(testSavedSearch.id, null).getOrNull()
    assertNull(result)
  }

  @Test
  fun `test find registered job list`() {
    repo.save(savedList.apply { sfJobOpp = testSFJobOpp })
    val results = repo.findRegisteredJobList(savedList.sfJobOpp?.sfId ?: fail()).getOrNull()
    assertNotNull(results)
    assertEquals(savedList.id, results.id)
  }

  @Test
  fun `test find registered job list fail not registered`() {
    repo.save(
      savedList.apply {
        sfJobOpp = testSFJobOpp
        registeredJob = false
      }
    )

    val results = repo.findRegisteredJobList("${testSFJobOpp.id}").getOrNull()
    assertNull(results)
  }

  @Test
  fun `test find registered job list fail id`() {
    repo.save(savedList.apply { sfJobOpp = testSFJobOpp })
    // This is unusual to have a string for the sfJobLink
    val results = repo.findRegisteredJobList("").getOrNull()
    assertNull(results)
  }

  @Test
  fun `test find short name ignore case`() {
    val result = repo.findByShortNameIgnoreCase(savedList.tbbShortName).getOrNull()
    assertNotNull(result)
    assertEquals(savedList.id, result.id)
  }

  @Test
  fun `test find short name ignore case fail`() {
    val result = repo.findByShortNameIgnoreCase("BOB").getOrNull()
    assertNull(result)
  }

  @Test
  fun `test find lists with jobs`() {
    repo.save(savedList.apply { sfJobOpp = testSFJobOpp })
    val result = repo.findListsWithJobs()
    assertNotNull(result)
    assertTrue { result.isNotEmpty() }
    assertEquals(savedList.id, result.first().id)
  }

  @Test
  fun `test find lists with jobs fail`() {
    repo.save(savedList.apply { status = Status.deleted })
    val result = repo.findListsWithJobs()
    assertNotNull(result)
    assertTrue { result.isEmpty() }
  }

  @Test
  fun `test find lists with jobs fail job opp null`() {
    val result = repo.findListsWithJobs()
    assertNotNull(result)
    assertTrue { result.isEmpty() }
  }
}
